finality-aleph/src/validator_network/io.rs

```diff
@@ -0,0 +1,176 @@
+}
+
+/// Attempts to receive some data using the stream.
+pub async fn receive_data<S: AsyncReadExt + Unpin, D: Data>(
+    mut stream: S,
+) -> Result<(S, D), ReceiveError> {
+    let mut buf = [0; 4];
+    stream
+        .read_exact(&mut buf[..])
+        .await
+        .map_err(|_| Error::ConnectionClosed)?;
+    let len = u32::from_le_bytes(buf);
+    if len > MAX_DATA_SIZE {
+        return Err(Error::DataTooLong(len).into());
+    }
+    let mut buf: Vec<u8> = vec![0; len as usize];
+    stream
+        .read_exact(&mut buf[..])
+        .await
+        .map_err(|_| Error::ConnectionClosed)?;
+    let data = D::decode(&mut &buf[..]).map_err(|_| ReceiveError::DataCorrupted)?;
+    Ok((stream, data))
+}
+
+#[cfg(test)]
+mod tests {
+    use tokio::io::{duplex, AsyncWriteExt};
+
+    use super::{receive_data, send_data, Error, ReceiveError, SendError, MAX_DATA_SIZE};
+
+    #[tokio::test]
+    async fn sends_and_receives_correct_data() {
+        let (sender, receiver) = duplex(4096);
+        let data: Vec<i32> = vec![4, 3, 43];
+        let _sender = send_data(sender, data.clone())
+            .await
+            .expect("data should send");
+        let (_receiver, received_data) = receive_data(receiver).await.expect("should receive data");
+        let received_data: Vec<i32> = received_data;
+        assert_eq!(data, received_data);
+    }
+
+    #[tokio::test]
+    async fn fails_to_receive_from_dropped_connection() {
+        let (_, receiver) = duplex(4096);
+        match receive_data::<_, i32>(receiver).await {
+            Err(e) => assert_eq!(ReceiveError::Error(Error::ConnectionClosed), e),
+            _ => panic!("received data from a dropped stream!"),
+        }
+    }
+
+    #[tokio::test]
+    async fn fails_to_send_to_dropped_connection() {
+        let (sender, _) = duplex(4096);
+        let data: Vec<i32> = vec![4, 3, 43];
+        match send_data(sender, data.clone()).await {
+            Err(e) => assert_eq!(SendError(Error::ConnectionClosed), e),
+            _ => panic!("send data to a dropped stream!"),
+        }
+    }
+
+    #[tokio::test]
+    async fn fails_to_receive_too_much_data() {
+        let (mut sender, receiver) = duplex(4096);
+        let too_long = MAX_DATA_SIZE + 43;
+        let payload = too_long.to_le_bytes().to_vec();
+        sender
+            .write_all(&payload)
+            .await
+            .expect("sending should work");
+        match receive_data::<_, i32>(receiver).await {
+            Err(e) => assert_eq!(ReceiveError::Error(Error::DataTooLong(too_long)), e),
+            _ => panic!("received too long data!"),
+        }
+    }
+
+    #[tokio::test]
+    async fn fails_to_decode_empty_data() {
+        let (mut sender, receiver) = duplex(4096);
+        let payload = 0u32.to_le_bytes().to_vec();
+        sender
+            .write_all(&payload)
+            .await
+            .expect("sending should work");
+        match receive_data::<_, i32>(receiver).await {
```

Comment:
Are you absolutely sure there are no valid objects that encode to an empty byte vec? Because this kind of assumes it is the case. I imagine that an empty array could probably encode to `vec![]`.

Reply:
Yup, but in this case we are explicitly using `i32` as the target type, and this does not encode into empty byte vecs.

Reply:
Oh, right, I got confused, I think it is correct. Will resolve in a minute after I convince myself :P